### PR TITLE
Syslog parser drop invalid

### PR DESCRIPTION
**Pull request by bazsi**
This PR contains a combination of a long-needed msg-format related refactor (moving common code out of format modules into the msg-format layer) and the ability for syslog-parser() to drop incorrectly formatted messages (right now it is only RFC5424 that can indicate invalid messages, RFC3164 consumes everything).

The motivation behind this feature is that more and more rfc5424 style messages are actually incorrectly formatted, while still containing the initial "1 " version indicator right after the priority field. Instead of making the parser less strict, the drop-invalid() setting allows the easy identification of these messages, so that an alternative parser can be applied, like this:

```
@version: 3.30

log {
        source { tcp(port(2000) flags(no-parse)); };

        if {
                parser { syslog-parser(drop-invalid(yes) flags(syslog-protocol)); };
        } else {
                # drop RFC5424 indicator as it was not right
                rewrite { subst("^(<[0-9]+>)1 ", "$1"); };
                parser { syslog-parser(drop-invalid(no)); };
        };
        destination { file("log"); };
};
```

This was a long requested feature by the https://github.com/splunk/splunk-connect-for-syslog team @rfaircloth-splunk

### CODE REVIEW

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by bazsi** [comment]
dropping "WIP", this should be good enough for review. The changes are not complex, however there's a series of them. I'd recommend going patch-by-patch.

The motiviation behind all the changes, the feature itself is trivial, once all the preparation is there.

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by mitzkia** [comment]
@kira-syslogng test this please;

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by rfaircloth-splunk** [comment]
looking forward to having this in the toolbox

File: lib/msg-format.c
Comment (gaborznagy) [inline lib/msg-format.c]: In case of `no-parse` flag, we still call `msg_format_postprocess_message`, and thus setting additional flags from config (e.g. assume-utf8, local).
I'm okay with this, but previously neither `syslog-format`, nor `linux-kmsg` set these flags.

File: lib/msg-format.c
Comment (gaborznagy) [inline lib/msg-format.c:144]: Why do we need to post-process messages that are failed to parse?
Isn't processing `no-multi-line` could interfere with the `>@<` mark from the error-injecting function?

I'm okay that `msg_format_parse_conditional` does post-processing, as the message will be passed over for further processing in the log path.

File: lib/msg-format.c
Comment (gaborznagy) [inline lib/msg-format.c:93]: I would move both `local` and `assume-utf8` flags to preprocess.
Reasons:
- syslog format handler in legacy-hander uses the `LF_LOCAL` flag
- regardles of `no-parse` flag, the post-processing is done unconditionally, so it doesn't matter if it's in pre- or post-processing.

File: modules/syslogformat/syslog-parser.c
Comment (gaborznagy) [inline modules/syslogformat/syslog-parser.c:47]: Wouldn't improve things a `G_UNLIKELY` here? Just like we do in `LogParser` with template option

**Review by gaborznagy — COMMENTED** [review]
I only have some questions/comments.

The new feature looks good to me!

File: lib/msg-format.c
Comment (gaborznagy) [reply lib/msg-format.c:93]: I think this is the reason that kira fails, as one test with flag `kernel` fails.

File: lib/msg-format.c
Comment (bazsi) [reply lib/msg-format.c:79]: I think it makes better sense that we apply flags to both parsed/unparsed/error cases. If we asked syslog-ng not to give us newlines in messages, that's possibly because we  have something in our processing pipeline that doesn't handle that.

If in rare cases we don't perform as asked, then that's not that intuitive.

So I think it's better to fix this behavior.

File: lib/msg-format.c
Comment (bazsi) [reply lib/msg-format.c:137]: I'll fix this.

File: lib/msg-format.c
Comment (bazsi) [reply lib/msg-format.c]: The behavioral change was intentional, and again it is the "should have been this way" case.

The utf8 validation/sanitization is not yet centralized (but probably should be), I didn't touch that because of the performance implications.

I could easily centralize assume-utf8, so I did that. And the message can be utf-8 even if it wasn't parsed.

The local flag was set even previously depending on the LP_LOCAL flag for the source, which is done by the file() source.

File: lib/msg-format.c
Comment (bazsi) [reply lib/msg-format.c:144]: The same reason as above, failed parsing results in a message too.

File: lib/msg-format.c
Comment (bazsi) [reply lib/msg-format.c:93]: Can you elaborate that?

File: modules/syslogformat/syslog-parser.c
Comment (bazsi) [reply modules/syslogformat/syslog-parser.c:47]: I can add this.

Comment (bazsi) [reply modules/syslogformat/syslog-parser.c:47]: Hmm.. I have second thoughts on this. A config may easily use drop_invalid mode exclusively, in which case G_UNLIKELY is not true.

In the case of the "template" option, that branch is slow anyway, so the G_UNLIKELY is the least of the concerns anyway.

So I would just leave it to the compiler.

File: lib/msg-format.c
Comment (bazsi) [reply lib/msg-format.c:97]: hmmm. I have now pushed a WIP patch on top, which clears all the flags in log_msg_clear(). I still have to give some more thoughts on the behavior, but that means that the error message becomes completely uncoupled with the original one.

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by bazsi** [comment]
@kira-syslogng retest this please;

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by gaborznagy** [comment]
@kira-syslogng retest this please branch=syslog-parser-update-wrong-format-test;

File: modules/syslogformat/syslog-parser.c
Comment (gaborznagy) [reply modules/syslogformat/syslog-parser.c:47]: I see, so this wouldn't make too much sense then.

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by gaborznagy** [comment]
I haven't checked why does the `test_secure_logging` UT fail only with cmake.
I've updated a test that was affected by the injected error message format, so kira is green now.

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by gaborznagy** [comment]
@kira-syslogng retest this please branch=syslog-parser-update-wrong-format-test;

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by gaborznagy** [comment]
@kira-syslogng retest this please branch=syslog-parser-update-wrong-format-test;

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by gaborznagy** [comment]
(oops, forgot to rebase the test branch, I will restart the kira test)

**Comment by gaborznagy** [comment]
@kira-syslogng retest this please branch=syslog-parser-update-wrong-format-test;

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by bazsi** [comment]
@kira-syslogng retest this please branch=syslog-parser-update-wrong-format-test;

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by bazsi** [comment]
@kira-syslogng retest this please branch=syslog-parser-update-wrong-format-test;

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by bazsi** [comment]
@kira-syslogng retest this please branch=syslog-parser-update-wrong-format-test;

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Review by Kokan — APPROVED** [review]

**Comment by Kokan** [comment]
Also if possible please write a news entry under `news/` directory. But this does not block the merge, if somebody merges this as is, they can write one and open a PR.

**Comment by bazsi** [comment]
added news file + rebased

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by bazsi** [comment]
@kira-syslogng retest this please branch=syslog-parser-update-wrong-format-test;

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Review by Kokan — APPROVED** [review]

**Review by gaborznagy — APPROVED** [review]

**Comment by gaborznagy** [comment]
I've created an internal doc update ticket about the new option and the changed behaviour of msg-format flags (i.e. from now on if "no-parse" flag is used additional flags e.g. "no-multi-line" can be used as well.)

**Comment by gaborznagy** [comment]
@bazsi I wonder: did we intentionally not document the injected error message?
This is the only thing in the documentation: 
> If syslog-ng OSE cannot parse a message, it results in an error

I would mention the details of the error message (rewritten program and message fields, modified facility and severity values).

**Comment by bazsi** [comment]
There was no intention, it just happened this way.

Documenting the error handling might indeed be useful, as well as the
drop-invalid option in the context of syslog-parser ().

**Comment by gaborznagy** [comment]
Thanks, I've submitted a ticket to the doc team.

**Pull request merged**
